tests/rest/media/v1/test_media_storage.py

```diff
@@ -201,36 +202,49 @@ class _TestImage:
     ],
 )
 class MediaRepoTests(unittest.HomeserverTestCase):
-
+    test_image: ClassVar[_TestImage]
     hijack_auth = True
     user_id = "@test:user"
 
     def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
 
-        self.fetches = []
+        self.fetches: List[
+            Tuple[
+                "Deferred[Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]]]",
+                str,
+                str,
+                Optional[QueryParams],
+            ]
+        ] = []
 
         def get_file(
             destination: str,
             path: str,
             output_stream: BinaryIO,
-            args: Optional[Dict[str, Union[str, List[str]]]] = None,
+            args: Optional[QueryParams] = None,
+            retry_on_dns_fail: bool = True,
             max_size: Optional[int] = None,
-        ) -> Deferred:
-            """
-            Returns tuple[int,dict,str,int] of file length, response headers,
-            absolute URI, and response code.
-            """
+            ignore_backoff: bool = False,
+        ) -> Deferred[Tuple[int, Dict[bytes, List[bytes]]]]:
+            """A mock for MatrixFederationHttpClient.get_file."""
 
-            def write_to(r):
+            def write_to(
+                r: Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]]
+            ) -> Tuple[int, Dict[bytes, List[bytes]]]:
                 data, response = r
                 output_stream.write(data)
                 return response
 
-            d = Deferred()
-            d.addCallback(write_to)
+            d: Deferred[Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]]] = Deferred()
+            # Callbacks can _change_ the value held within a resolved Deferred(!!).
+            # The return type of addCallback reflects this...
+            d_after_callback = d.addCallback(write_to)
+            # ... even though d.addCallback returns d!
+            assert d_after_callback is d  # type: ignore[comparison-overlap]
```

Comment:
Brace yourself.

The `assert d_after_callback is d` is me sanity checking that I've understood what is going on. It can be removed, particularly if we find it actively worsens clarity.

Taking a step back, there were two complaints. The first is that `d` needs an annotation describing what it holds.
My reasoning:

1. `d: Deferred[T]` for some type `T`. I have always understood `T` to mean the type that `d` "holds" after it resolves successfully, and the type that is passed to its callback. To determine `T`, we need to find where `d` is resolved.
2. Where is `d` used?
  - It is appended to self.fetches on the line above: https://github.com/matrix-org/synapse/blob/7b67e93d499cb45f7217e9dfea046ed8b5c455fd/tests/rest/media/v1/test_media_storage.py#L231
   and later resolved with a value here https://github.com/matrix-org/synapse/blob/7b67e93d499cb45f7217e9dfea046ed8b5c455fd/tests/rest/media/v1/test_media_storage.py#L301-L303 and here. This suggests that `T` is of type `T1 = Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]`.
  - It is returned by `get_file`: https://github.com/matrix-org/synapse/blob/7b67e93d499cb45f7217e9dfea046ed8b5c455fd/tests/rest/media/v1/test_media_storage.py#L232
   where it is presumably used by Synapse. This suggests that `T` is the return type of `MatrixFederationHttpClient.get_file`, namely `T2 = Tuple[int, Dict[bytes, List[bytes]]`.
3. The two return types are different. The callbacks on `d` must be responsible for fixing this up. There is only one callback, `write_to`. That suggests that `write_to: Callable[T1, T2]`.
4. To keep mypy happy, we need to return `Deferred[T2]` from `get_file`. But we only have `d: Deferred[T1]`. Fortunately, the expression `d.addCallback(write_to)` is of type `Deferred[T2]` (though it is the same underlying object as `d`). (Note that Twisted [has an explicit cast](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/defer.py#L514-L516) for this.)
---

Underlying all this, I think there is some friction between how Deferreds are used and how their type annotations mean they're interpreted by mypy. (See also https://github.com/twisted/twisted/issues/11772)

A `Deferred` is considered an Awaitable[`_DeferredResultT`], where `_DeferredResultT` is contravariant. This is somewhat confusing.

- If we think of Deferred[T] as "eventually T" (i.e. some kind of future) then T ought to be covariant (see e.g. [typeshed's asyncio.Future](https://github.com/python/typeshed/blob/main/stdlib/asyncio/futures.pyi#L25-L32).
- If we think of Deferred[T] as "a callback which will be given a T", then T ought to be contravariant, as it is today.

I wonder if it might make more sense to consider Deferred as generic over two parameters: a contravariant callback-arg type and a covariant return type. I.e. closer to Callable[T, R] than Future[T].

Additionally: I think there is a potential footgun here if we keep references around to the old. I think it's best illustrated with an example:

```python
def ab1(a: A) -> B: ...
def ab2(a: A) -> B: ...

# DANGER!
d: Deferred[A] = Deferred()
d.addCallback(ab1)  # okay
d.addCallback(ab2)  # not okay, but accepted by the typechecker

# Safety
Deferred().addCallback(ab1)  # okay, of type Deferred[B]
Deferred().addCallback(ab1).addCallback(ab2)  # not okay, and rejected by the typechecker
```

(If the type system allowed `addCallback` to fully consume `d`---think a Rust method taking `self` rather than `&mut self`---the `DANGER!` section would be forbidden at build time. But we don't live in that world.)

Reply:
Right -- I think the explict assert makes it more confusing (and is somewhat testing twisted behavior, although I don't know if `Deferred` guarantees that or not).

Reply:
e22b4e2 tidies this up a bit.